[KEP]Add job interface for integration (#537)
kubernetes-sigs/kueue: keps/369-job-interface/README.md

    - [Unit Tests](#unit-tests)
    - [Integration tests](#integration-tests)
  - [Graduation Criteria](#graduation-criteria)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Summary

<!--
This section is incredibly important for producing high-quality, user-focused
documentation such as release notes or a development roadmap. It should be
possible to collect this information before implementation begins, in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself. KEP editors and SIG Docs
should help to ensure that the tone and content of the `Summary` section is
useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

> Define a set of methods(known as golang interface) to help other job-like applications to smoothly integrate with Kueue.

Author: Added

and make progress.
-->

- Integrate with other job-like applications

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

We collected feedbacks from the community about how to fully integrate Kueue with MPIJob,
see [#499](https://github.com/kubernetes-sigs/kueue/issues/499).

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

> - Job interface here is more like a template and tell developers what to do, but even without

Reviewer: To be honest I'm lost here. Is it a template, a hint, a set of best practices or a library?

implementing the interface people can also build their controllers.
- This will increase the code complexity by wrapping the original Jobs.

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

> Part I: Job Interface

Reviewer: What is missing (or at least I'm) missing is the overall mechanic. Who is calling what and when? What is the flow?

Reviewer 2: ```suggestion
### Job Interface
```

We will define a new interface named Job, this should be implemented by custom job-like applications:

```golang
type Job interface {
  // Suspend instructs whether the job is suspended or not.
  Suspend() bool
  // Start means starting the job, like unsuspending the job.

>   Start(ctx context.Context, client client.Client, wl *kueue.Workload) error

Reviewer 2: I was thinking that maybe it was about nodeAffinity, but you have separate methods for that.

>   Start(ctx context.Context, client client.Client, wl *kueue.Workload) error

Author: Start is not good designed I think, changed to:
```
IsSuspend() bool        --> former is Suspend() bool
Suspend() error         --> former is Start() error
Unsuspend() error       --> former is Stop() error
```

We'll have a `Start()` helper function to call `Unsuspend()` and `InjectNodeAffinity` separately.

  // Stop means stopping the job, like suspending the job.
  Stop(ctx context.Context, client client.Client, wl *kueue.Workload) error
  // InjectNodeAffinity will inject the node affinity extracting from workload to job.
  InjectNodeAffinity(ctx context.Context, client client.Client, wl *kueue.Workload) error
  // RestoreNodeAffinity will restore the original node affinity of job.
  RestoreNodeAffinity(ctx context.Context, client client.Client, wl *kueue.Workload) error
  // Finished instructs whether the job is completed/failed or not.
  Finished() bool
  // Finish will update the workload to finished.

>   Finish(ctx context.Context, client client.Client, wl *kueue.Workload) error

Reviewer 2: Why does the Job need to implement this? This should be part of the helper code.

  // EquivalentToWorkload validates whether the workload is semantically equal to the job.

>   EquivalentToWorkload(wl kueue.Workload) bool

Reviewer 2: Can you remind me where we use this?

Reviewer 2: I guess it makes sense to keep it for now. I think currently if you change the parallelism of a running job, we would delete the Workload object to cause it to be requeued.

Author: Yes, this is needed when job can be modified in runtime.

  // CreateWorkload will create a workload constructing by the job.

>   CreateWorkload(ctx context.Context, client client.Client, scheme *runtime.Scheme) (*kueue.Workload, error)

Reviewer 2: Maybe this should be something like `PodSets` that will return `[]kueue.PodSet`. Then from there we can provide code to create the Workload object.

Reviewer 2: We can for sure obtain the namespace from the ObjectMeta. I wonder if there are any APIs that already include a queue name in a field. Otherwise we can also obtain the queue name from the (standard) annotation in the ObjectMeta.

Author: Fine, then we'll have a help function named like `createWorkload()` with client passed-in.

  // Priority returns the job's priority

>   Priority(ctx context.Context, client client.Client) (string, int32, error)

Reviewer 2: `PriorityClass`? And then we can have code to get the priority from the client. Although maybe some custom jobs might choose to give you a number already.

  // QueueName returns the queue name the job enqueued.
  QueueName() string
  // Ignored instructs whether this job should be ignored in reconciling, e.g. lacking the queueName.
  Ignored() bool
  // PodsCount returns the pods number job requires.
  PodsCount() int32
}
```

Also defined another interface for all-or-nothing scheduling required job-like applications:

```golang
type GangSchedulingJob interface {
  // GangSchedulingSucceeded instructs whether job is scheduled successfully in gang-scheduling.

>   GangSchedulingSucceeded() bool

Reviewer 3: `GangSchedulingJob` doesn't sound appropriate to me. Since we don't use the expression `GangSchedulling` in the document and kueue doesn't guarantee gang-schedule all Pods to Nodes.

https://github.com/kubernetes-sigs/kueue/blob/160021fcd342f0b010e4d0d5e64ea634dba54df2/docs/tasks/setup_sequential_admission.md

However, I'm not sure what name is appropriate.

>   GangSchedulingSucceeded() bool

Reviewer 2: waitForPodsReady is currently a Kueue level configuration, so it doesn't make sense to have it as part of the GenericJobInterface for now.

>   GangSchedulingSucceeded() bool

Reviewer 3: > waitForPodsReady is currently a Kueue level configuration, so it doesn't make sense to have it as part of the GenericJobInterface for now.

I agree. I wanted to ask whether `Gang` is a general term or not.

>   GangSchedulingSucceeded() bool

Author: My original thought is this may also need by other jobs, by implementing like https://github.com/kubernetes-sigs/kueue/blob/3268949a656a18b779c6d5369de8a1a92034cef4/pkg/controller/workload/job/job_controller.go#L255-L258

But we can extend the interface if needed, removed then.

>   GangSchedulingSucceeded() bool

Reviewer 2: ah, I see what you mean now. Let's better be more explicit and make it `PodsReady() bool`

>   GangSchedulingSucceeded() bool

Reviewer 3: I see too. @alculquicondor's idea sounds good to me.

>   GangSchedulingSucceeded() bool

Author: Add a new interface SequenceJob to containing `PodsReady()`

>   GangSchedulingSucceeded() bool

Reviewer 2: Does it need to be a separate interface?

>   GangSchedulingSucceeded() bool

Author: Not every job needs this, I treat it as an advanced feature, so maybe it's good to keep the interface as small as possible.

>   GangSchedulingSucceeded() bool

Reviewer 2: What would we do with a job that doesn't implement the interface if the kueue is configured with `waitForReadyPods: true`?

the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

We collected feedbacks from the community about how to fully integrate Kueue with MPIJob,
see [#499](https://github.com/kubernetes-sigs/kueue/issues/499).

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

- Job interface defined here is a hint for developers to build their own controllers,
it's not a hard constraint(to implement the interface) but a good reference.
- This will increase the code complexity by wrapping the original Jobs.

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

> ## Design Details

Reviewer 3: Can we also provide interfaces for webhook like https://github.com/kubernetes-sigs/kueue/blob/160021fcd342f0b010e4d0d5e64ea634dba54df2/pkg/controller/workload/job/job_webhook.go?

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

Part I: Job Interface

We will define a new interface named GenericJob, this should be implemented by custom job-like applications:

```golang
type GenericJob interface {
  // Suspend instructs whether the job is suspended or not.
  Suspend() bool
  // Start will arm the job ready to start, like unsuspending the job and injecting nodeAffinity.

>   Start(ctx context.Context, wl *kueue.Workload) error

Reviewer 2: why pass a context? If there is no kubeclient, context is probably not necessary either?

Author: Yes, forgot to remove, same to other methods.

  // Stop will disarm the job ready to stop, like suspending the job and restoring nodeAffinity.
  Stop(ctx context.Context, wl *kueue.Workload) error
  // InjectNodeAffinity will inject the node affinity extracting from workload to job.
  InjectNodeAffinity(ctx context.Context, nodeSelectors []map[string]string) error
  // RestoreNodeAffinity will restore the original node affinity of job.
  RestoreNodeAffinity(ctx context.Context, nodeSelectors []map[string]string) error
  // Finished means whether the job is completed/failed or not.
  Finished() bool
  // ConstructWorkload will build a workload corresponding to the job.
  ConstructWorkload() (*kueue.Workload, error)
  // PriorityClass returns the job's priority class name.
  PriorityClass() string
  // QueueName returns the queue name the job enqueued.
  QueueName() string
  // Ignored instructs whether this job should be ignored in reconciling, e.g. lacking the queueName.
  Ignored() bool
  // PodsCount returns the pods number job requires.

>   PodsCount() int32

Reviewer 2: Should this return an array? Why do we need this in addition to ConstructWorkload?

Author: ConstructWorkload removed now.

Author: Yes, pushed.

this helps us to build a multi-tenant job queueing system in Kubernetes, this is attractive to other job-like applications
like MPIJob, who lacks the capacity of queueing.

The good news is Kueue is extensible and simple to integrate with through the intermediate medium, we named _Workload_ in Kueue,
what we need to do is to build a controller to reconcile the workload and the job itself.

But the complexity lays in developers who are familiar with job-like applications may have little knowledge of the
implementation details of Kueue and they have no idea where to start to build the controller. In this case, if we can provide an
interface which defines the default behaviors of the Job, and serve the Kubernetes Job as a standard template, it will do them
a great favor.

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

- Define an interface which shapes the default behaviors of Job
- Make Kubernetes Job an implementation template of the interface

> - Provide a library to organize helper functions which is used in constructing controllers

Reviewer 2: ```suggestion
- Provide a library to build integration controllers for job APIs that implement the interface
```

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

> - Integrate with other job-like applications

Reviewer 2: ```suggestion
- Integrate any job-like applications
```

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

> We collected feedbacks from the community about how to fully integrate Kueue with MPIJob,

Reviewer 2: ```suggestion
We collected feedback from the community about how to fully integrate Kueue with MPIJob,
```

see [#499](https://github.com/kubernetes-sigs/kueue/issues/499).

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

- Job interface defined here is a hint for developers to build their own controllers,

> it's not a hard constraint(to implement the interface) but a good reference.

Reviewer 2: It's a hard constrain if they wish to use the library, but they can always write the controllers from scratch

- This will increase the code complexity by wrapping the original Jobs.

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

Part I: Job Interface

We will define a new interface named GenericJob, this should be implemented by custom job-like applications:

```golang
type GenericJob interface {
  // Object returns the job instance.
  Object() client.Object
  // IsSuspend returns whether the job is suspend or not.

>   IsSuspend() bool

Reviewer 2: ```suggestion
  IsSuspended() bool
```

  // Suspend will suspend the job.

>   Suspend() error

Reviewer 4: 1) Preemption; 2) for some reason the job was unsuspended but was never admitted.

The expected behavior is for the job controller to delete all active pods.

  // UnSuspend will unsuspend the job.

>   UnSuspend() error

Reviewer: How will the library/controller know which job(and of what kind) to unsuspend because a particular workload got admitted? There is no key/identrifier returned by this Interface.

  // ResetStatus will reset the job status to the original state.
  // If true, status is modified, if not, status is as it was.

>   ResetStatus() bool

Reviewer 2: What is this for?

  // InjectNodeAffinity will inject the node affinity extracting from workload to job.

>   InjectNodeAffinity(nodeSelectors []map[string]string) error

Reviewer 2: ```suggestion
  InjectNodeSelectors(nodeSelectors []map[string]string) error
```

  // RestoreNodeAffinity will restore the original node affinity of job.

>   RestoreNodeAffinity(nodeSelectors []map[string]string) error

Reviewer 2: ```suggestion
  RestoreNodeSelectors(nodeSelectors []map[string]string) error
```

>   RestoreNodeAffinity(nodeSelectors []map[string]string) error

Reviewer: Why is this needed?

  // Finished means whether the job is completed/failed or not,
  // condition represents the workload finished condition.

>   Finished() (condition metav1.Condition, finished bool)

Author: Doubt about the return values of `condition`, but without it, we have to iterate the `j.Status.Conditions` again to get the condition. Or return the conditionType instead, similar to
https://github.com/kubernetes-sigs/kueue/blob/b6bdec3bfd7427bf69af74398ddcbae4acf35d76/pkg/controller/workload/job/job_controller.go#L511-L518

Author: This is more clear https://github.com/kubernetes-sigs/kueue/blob/ebf2b812705d0b82ce00fe3dbb39fc008d5a0d77/pkg/controller/workload/job/job_controller.go#L310-L320

Reviewer 2: Do we need the job condition or do we just need to know whether it finished?

Also, some APIs might not use metav1.Condition (kubeflow jobs don't).

>   Finished() (condition metav1.Condition, finished bool)

Reviewer: Do we expect that the user will actively query the object to get the state. Wouldn't a notification channel be more convenient?

>   // PodSets will build workload podSets corresponding to the job.

Reviewer 2: ```suggestion
  // PodSets the podSets corresponding to the job.
```

  PodSets() []kueue.PodSet
  // EquivalentToWorkload validates whether the workload is semantically equal to the job.

>   EquivalentToWorkload(wl kueue.Workload) bool

Reviewer: We probably need 2 methods instead of one:

* To tell what workload name matches to that particular GenericJob.
* To tell that workload and generic job got out of sync.

  // PriorityClass returns the job's priority class name.
  PriorityClass() string
  // QueueName returns the queue name the job enqueued.
  QueueName() string
  // Ignored instructs whether this job should be ignored in reconciling, e.g. lacking the queueName.

>   Ignored() bool

Reviewer 2: Maybe:

```suggestion
  IgnoreFromQueueing() bool
```

}
```

Also defined another interface `SequenceJob` for scheduling in sequence requirements, see design details here [All-or-nothing semantics for job resource assignment](https://github.com/kubernetes-sigs/kueue/tree/main/keps/349-all-or-nothing#kep-349-all-or-nothing-semantics-for-job-resource-assignment).

```golang
type SequenceJob interface {
    // PodsReady instructs whether job derived pods are all ready now.

>     PodsReady() bool

Reviewer 2: I feel like it should just be part of the main interface, and if they don't have any semantics for it, they should return true.

}
```

> Part II: Kubernetes Job implementation details

Reviewer 2: ```suggestion
### Kubernetes Job implementation details
```

We'll wrap the batchv1.Job to `BatchJob` who implements the GenericJob interface, as well as SequenceJob interface.

```golang

> type BatchJob struct {
>   batchv1.Job
> }

Reviewer 2: ```suggestion
type BatchJob  batchv1.Job
```

var _ GenericJob = &BatchJob{}
var _ SequenceJob = &BatchJob{}
```

Part III: Library

> We'll have several helper functions in constructing controllers, they can be imported directly in the future:

Reviewer 2: I think we can leave this to the code review. The pseudocode below should be enough

> We'll have several helper functions in constructing controllers, they can be imported directly in the future:

Reviewer: What about providing the full controller instead of just helper function? So that the user can just implement the interface and be done with it.

// StartJob will unsuspend the job, and also inject the node affinity.
StartJob(ctx context.Context, client client.Client, wl *kueue.Workload, condition metav1.Condition) error

// StopJob will suspend the job, and also restore node affinity, reset job status if needed.
 StopJob(ctx context.Context, client client.Client, job GenericJob, wl *kueue.Workload) error

 // CreateWorkload will derive and create a workload from the corresponding job.
 CreateWorkload(ctx context.Context, client client.Client, scheme *runtime.Scheme, job GenericJob) (*kueue.Workload, error)

 // GetNodeSelectors will extract node selectors from admitted workloads.
 GetNodeSelectors(ctx context.Context, client client.Client, w *kueue.Workload) ([]map[string]string, error)

 // UpdateQueueNameIfChanged will update workload queue name if changed.
 UpdateQueueNameIfChanged(ctx context.Context, client client.Client, job GenericJob, wl *kueue.Workload) error

 // SetWorkloadCondition will update the workload condition by the provide one.
 SetWorkloadCondition(ctx context.Context, client client.Client, wl *kueue.Workload, condition metav1.Condition) error
```

Part IV: Working Flow in pseudo-code

```golang

> Reconcile:

Reviewer 2: I think this should be part of the library. Something along these lines:

```golang
type Controller[J GenericJob] struct {}

func NewController[J GenericJob]() Controller[J] {
  return Controller[J]{}
}
```

Author: Some hooks maybe needed, but we can add them later.

Author: The disadvantage is the interface may be changed more frequently. We can make the interface as small as possible to mitigate this.

- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
- [Feature gate][feature gate] lifecycle
- [Deprecation policy][deprecation-policy]

[feature gate]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
-->

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

- 2023.01.09: KEP proposed for review, including motivation, proposal, risks,
test plan.

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

> - Developers who want to implement their own controllers will have to read more source code for

Reviewer 2: They will have to read less code than today, so it's not a drawback :)

better understandings.

> - Bad for the project long-term maintenance.

Reviewer 2: I wouldn't say bad, but it does add complexity: we should try to maintain the number of changes to the interface to a minimum

> - Bad for the project long-term maintenance.

Reviewer: I don't agree. I would say that such interface (together with library/controller boilerplate) is critical for Kueue. Who and how otherwise will integrate with it?

## Alternatives

<!--
What other approaches did you consider, and why did you rule them out? These do
not need to be as detailed as the proposal, but should include enough
information to express the idea and why it was not acceptable.

> -->

Reviewer 2: Each job implements their controller from scratch
